### PR TITLE
feat(dashboardsv2): Add world map chart

### PR DESCRIPTION
**Pull request by dashed**
Screen Shot 2021-01-20 at 1 05 54 PM

## TODO

- [x] restrict fields to pre-aggregates
- [x] depends on https://github.com/getsentry/sentry/pull/23199
- [x] add some tests

### CODE REVIEW

File: src/sentry/static/sentry/app/components/charts/worldMapChart.tsx [hunk 1]
```diff
@@ -124,7 +137,7 @@ class WorldMapChart extends React.Component<Props, State> {
     return (
       <BaseChart
         options={{
-          backgroundColor: theme.backgroundSecondary,
+          backgroundColor: theme.white,
```
Comment (Zylphrex) [inline src/sentry/static/sentry/app/components/charts/worldMapChart.tsx:140]: Won't this be problematic with dark mode? I think you want `theme.background` here.

Comment (dashed) [reply src/sentry/static/sentry/app/components/charts/worldMapChart.tsx]: I haven't audited Dashboards v2 in dark mode yet. But I've applied your suggestion.

File: src/sentry/static/sentry/app/views/dashboardsV2/widgetCard.tsx [hunk 2]
```diff
@@ -69,6 +70,193 @@ class WidgetCard extends React.Component<Props> {
...
+class WidgetCardVisuals extends React.Component<WidgetCardVisualsProps> {
+  shouldComponentUpdate(nextProps: WidgetCardVisualsProps): boolean {
+    // Widget title changes should not update the WidgetCardVisuals component tree
+    const currentProps = {
+      ...this.props,
+      widget: {
+        ...this.props.widget,
+        title: '',
+      },
+    };
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/dashboardsV2/widgetCard.tsx:247]: Would this be simpler as a `cloneDeep` operation? Something like

```ts
const currentProps = cloneDeep(this.props);
currentProps.widget.title = '';
```

File: src/sentry/static/sentry/app/views/dashboardsV2/widgetCard.tsx [hunk 3]
```diff
@@ -133,9 +316,74 @@ class WidgetCard extends React.Component<Props> {
...
     const {location, router, selection} = this.props;
     const {start, end, period} = selection.datetime;
 
+    if (widget.displayType === 'world_map') {
+      const DEFAULT_GEO_DATA = {
+        title: '',
+        data: [],
+      };
+
+      const processTableResults = () => {
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/dashboardsV2/widgetCard.tsx:336]: Should this be done in `WidgetQueries` it looks like you have a path there for worldmaps and having the request and result processing steps co-located would be good. That would leave this component focused on selecting the right rendering

File: src/sentry/static/sentry/app/views/dashboardsV2/widgetQueries.tsx [hunk 4]
```diff
@@ -164,7 +165,49 @@ class WidgetQueries extends React.Component<Props, State> {
...
+              loading: completed === promises.length ? false : true,
+            };
+          });
+        } catch (err) {
+          const errorMessage =
+            err?.responseJSON?.detail || t('An unknown error occurred.');
+          this.setState({errorMessage});
+        }
+      });
+    } else if (widget.displayType === 'world_map') {
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/dashboardsV2/widgetQueries.tsx:181]: This method will need a revisit soon its getting long and will have a fair amount of duplication when we add stat charts.

File: src/sentry/api/endpoints/organization_events.py [hunk 5]
```diff
@@ -126,3 +126,44 @@ def data_fn(offset, limit):
...
+        if not self.has_feature(organization, request):
+            return Response(status=404)
+
+        try:
+            params = self.get_snuba_params(request, organization)
+        except NoProjects:
+            return Response([])
+
+        fields = request.GET.getlist("field")
+        fields = ["geo.country_code", fields[0]]
```
Comment (Zylphrex) [inline src/sentry/api/endpoints/organization_events.py:142]: Should this just be
```suggestion
        fields = ["geo.country_code", request.GET.get("field")]
```
and add some basic validation on `field`.

File: src/sentry/api/endpoints/organization_events.py [hunk 6]
```diff
@@ -126,3 +126,44 @@ def data_fn(offset, limit):
...
+
+        def data_fn(offset, limit):
+            return discover.query(
+                selected_columns=fields,
+                query=request.GET.get("query"),
+                params=params,
+                orderby=self.get_orderby(request),
```
Comment (Zylphrex) [inline src/sentry/api/endpoints/organization_events.py:149]: Does this need an user defined order by?

File: src/sentry/api/endpoints/organization_events.py [hunk 7]
```diff
@@ -126,3 +126,44 @@ def data_fn(offset, limit):
...
+                offset=offset,
+                limit=limit,
+                referrer=request.GET.get("referrer", "api.organization-events-geo"),
+                auto_fields=True,
+                auto_aggregations=True,
+                use_aggregate_conditions=True,
```
Comment (Zylphrex) [inline src/sentry/api/endpoints/organization_events.py:155]: Do we want these to be `True`? Do you see the user making queries like give me the `p75(transaction.duration)` by country and then filtering out countries where the `count():>100`?

File: src/sentry/api/endpoints/organization_events.py (hunk 6 above)
Comment (dashed) [reply src/sentry/api/endpoints/organization_events.py:148]: No. I'll remove this.

File: src/sentry/api/endpoints/organization_events.py (hunk 7 above)
Comment (dashed) [reply src/sentry/api/endpoints/organization_events.py]: I removed `auto_fields` and `auto_aggregations`. 

Is it fine if `use_aggregate_conditions` is kept?

File: src/sentry/static/sentry/app/views/dashboardsV2/widgetCard.tsx (hunk 2 above)
Comment (dashed) [reply src/sentry/static/sentry/app/views/dashboardsV2/widgetCard.tsx]: That might be more costly than a simple shallowclone. I don't intend to overwrite anything in the props.

File: src/sentry/api/endpoints/organization_events.py [hunk 8]
```diff
@@ -126,3 +127,48 @@ def data_fn(offset, limit):
...
+        except NoProjects:
+            return Response([])
+
+        maybe_aggregate = request.GET.get("field")
+
+        if not maybe_aggregate:
+            return Response({"detail": ["No column selected"]}, status=400)
+
+        if not is_function(maybe_aggregate):
+            return Response({"detail": ["Functions may only be given"]}, status=400)
```
Comment (dashed) [inline src/sentry/api/endpoints/organization_events.py:148]: @Zylphrex is this good enough?

Comment (markstory) [reply src/sentry/api/endpoints/organization_events.py:148]: API error messages are usually strings.

File: src/sentry/api/endpoints/organization_events.py [hunk 9]
```diff
@@ -126,3 +127,46 @@ def data_fn(offset, limit):
...
+
+        try:
+            params = self.get_snuba_params(request, organization)
+        except NoProjects:
+            return Response([])
+
+        maybe_aggregate = request.GET.get("field")
+
+        if not maybe_aggregate:
+            return Response({"detail": ["No column selected"]}, status=400)
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_events.py:145]:
```suggestion
            raise ParseError(detail="No column selected")
```

You'd also need `from rest_framework.exceptions import ParseError` in the import list.

File: src/sentry/api/endpoints/organization_events.py [hunk 10]
```diff
@@ -126,3 +127,46 @@ def data_fn(offset, limit):
...
+
+        if not is_function(maybe_aggregate):
+            return Response({"detail": ["Functions may only be given"]}, status=400)
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_events.py:148]:
```suggestion
            raise ParseError(detail="Functions may only be given")
```

File: src/sentry/static/sentry/app/components/dashboards/widgetQueryFields.tsx [hunk 11]
```diff
@@ -100,6 +103,12 @@ function WidgetQueryFields({displayType, errors, fields, fieldOptions, onChange}
             fieldValue={explodeField({field})}
             fieldOptions={fieldOptions}
             onChange={value => handleChangeField(value, i)}
+            filterPrimaryOptions={option => {
+              if (option.value.kind === FieldValueKind.FUNCTION) {
+                return true;
+              }
+              return false;
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/dashboards/widgetQueryFields.tsx:110]:
```suggestion
              return option.value.kind === FieldValueKind.FUNCTION;
```
🏌️

File: src/sentry/static/sentry/app/components/modals/addDashboardWidgetModal.tsx [hunk 12]
```diff
@@ -159,6 +160,36 @@ class AddDashboardWidgetModal extends React.Component<Props, State> {
...
+        let newQueries = prevState.queries;
+
+        // Filter out non-aggregate fields
+        newQueries = newQueries.map(query => {
+          const fields = query.fields.filter(isAggregateField);
+          return {
+            ...query,
+            fields: fields.length ? fields : ['count()'],
+          };
+        });
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/modals/addDashboardWidgetModal.tsx:178]: Should we have a test for this?

**Review by markstory — APPROVED** [review]
Looks good we can refactor WidgetQueries when the stat widget is added.

**Review by markstory — APPROVED** [review]

**Pull request merged**